torch/distributed/rpc/rref_proxy.py

```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
 def _local_invoke_async_execution(rref, func_name, args, kwargs):
     return getattr(rref.local_value(), func_name)(*args, **kwargs)
 
+# convert a Future[Future[X]] to a Future[X]
+def _unwrap_future(fut: Future) -> Future:
+    res: Future = Future()
+
+    def _unwrap_level0(fut0):
+        try:
+            res.set_result(fut0.wait())
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    def _unwrap_level1(fut1):
+        try:
+            fut1.wait().then(_unwrap_level0)
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    fut.then(_unwrap_level1)
+    return res
+
 def _invoke_rpc(rref, rpc_api, func_name, timeout, *args, **kwargs):
-    # Since rref._get_type can potentially issue an RPC, it should respect the
-    # passed in timeout here.
-    rref_type = rref._get_type(timeout=timeout)
-
-    _invoke_func = _local_invoke
-    # Bypass ScriptModules when checking for async function attribute.
-    bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    # this function returns either the value or a Future depending on what rpc_api is.
+    def _rref_type_cont(future, is_future):
+        if is_future:
+            rref_type = future.wait()
+        else:
+            rref_type = future
+
+        _invoke_func = _local_invoke
+        # Bypass ScriptModules when checking for async function attribute.
+        bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
+            rref_type, torch._C.ScriptModule
+        )
+        if not bypass_type:
+            func = getattr(rref_type, func_name)
+            if hasattr(func, "_wrapped_async_rpc_function"):
+                _invoke_func = _local_invoke_async_execution
+
+        return rpc_api(
+            rref.owner(),
+            _invoke_func,
+            args=(rref, func_name, args, kwargs),
+            timeout=timeout
+        )
+
+    request_blocking = rpc_api != rpc_async
+    rref_val = rref._get_type(timeout=timeout, blocking=request_blocking)
```

Comment:
It might be easier to make `_get_type` always return a future to reduce special casing here (and in other callers of this API?) If so, we can feel free to make this change in a follow up PR.

Reply:
Yeah, always returning a future would eliminate most of the complexity here. I'll adjust this change to assume this will be the future behavior by always calling with `blocking=False`.